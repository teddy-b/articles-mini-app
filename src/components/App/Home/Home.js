import React, { Component } from 'react';
import { PropTypes } from 'prop-types';

import ArticlesDetails from './ArticlesDetails';
import ArticlesList from './ArticlesList';
import ErrorBoundary from '../../shared/ErrorBoundary';
import Loading from '../../shared/Loading';

import './Home.css';

class Home extends Component {
  componentDidMount() {
    const {
      match: { params: { id } },
      onLoadArticles,
      onLoadComments,
      onSelectArticle
    } = this.props;

    onLoadArticles();

    if (id) {
      const articleId = parseInt(id, 10);
      onSelectArticle(articleId);
      onLoadComments(articleId);
    }
  }

  render() {
    const {
      articles,
      articlesCount,
      comments,
      fetching,
      onAddComment,
      onAddReply,
      onLoadComments,
      onLoadReplies,
      onNextPage,
      onPrevPage,
      onSelectArticle,
      page,
      selectedArticleId
    } = this.props;
    const selectedArticle = articles.filter(a => a.id === selectedArticleId)[0];

    // const { match: { params: { id } } } = this.props;

    return (
      <div className="home">
        <ErrorBoundary>
          {fetching && <Loading />}
        </ErrorBoundary>
        <ErrorBoundary>
          <ArticlesList
            articles={articles}
            articlesCount={articlesCount}
            onLoadComments={onLoadComments}
            onNextPage={onNextPage}
            onPrevPage={onPrevPage}
            onSelectArticle={onSelectArticle}
            page={page}
            selectedArticleId={selectedArticleId}
          />
        </ErrorBoundary>
        <ErrorBoundary>
          <ArticlesDetails
            comments={comments}
            onAddComment={onAddComment}
            onAddReply={onAddReply}
            onLoadReplies={onLoadReplies}
            selectedArticle={selectedArticle}
          />
        </ErrorBoundary>
      </div>
    );
  }
}

Home.propTypes = {
  articles: PropTypes.array,
  articlesCount: PropTypes.number,
  comments: PropTypes.array,
  fetching: PropTypes.bool,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string
    })
  }),
  onAddComment: PropTypes.func,
  onAddReply: PropTypes.func,
  onLoadArticles: PropTypes.func,
  onLoadComments: PropTypes.func,
  onLoadReplies: PropTypes.func,
  onNextPage: PropTypes.func,
  onPrevPage: PropTypes.func,
  onSelectArticle: PropTypes.func,
  page: PropTypes.number,
  selectedArticleId: PropTypes.number
};

export default Home;
