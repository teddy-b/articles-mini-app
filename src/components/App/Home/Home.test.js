import React from 'react';
import renderer from 'react-test-renderer';

import Home from './Home';

jest.mock('./ArticlesDetails', () => 'ArticlesDetails');
jest.mock('./ArticlesList', () => 'ArticlesList');
jest.mock('../../shared/ErrorBoundary', () => 'ErrorBoundary');
jest.mock('../../shared/Loading', () => 'Loading');

describe('Home', () => {
  const onLoadCommentsMock = jest.fn();
  const props = {
    articles: [],
    articlesCount: 0,
    comments: [],
    fetching: false,
    match: {
      params: {
      }
    },
    onAddComment: jest.fn(),
    onAddReply: jest.fn(),
    onLoadArticles: jest.fn(),
    onLoadComments: onLoadCommentsMock,
    onLoadReplies: jest.fn(),
    onNextPage: jest.fn(),
    onPrevPage: jest.fn(),
    onSelectArticle: jest.fn(),
    page: 0,
    selectedArticleId: null
  };

  it('renders correctly', () => {
    const component = renderer.create(<Home {...props} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with params', () => {
    const matchMock = {
      params: {
        id: '1'
      }
    };
    const component = renderer.create(
      <Home
        {...props}
        match={matchMock}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly while fetching', () => {
    const fetching = true;
    const component = renderer.create(
      <Home
        {...props}
        fetching={fetching}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
