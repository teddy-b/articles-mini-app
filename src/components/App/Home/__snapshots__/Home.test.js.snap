// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<div
  className="home"
>
  <ErrorBoundary />
  <ErrorBoundary>
    <ArticlesList
      articles={Array []}
      articlesCount={0}
      onLoadComments={[Function]}
      onNextPage={[Function]}
      onPrevPage={[Function]}
      onSelectArticle={[Function]}
      page={0}
      selectedArticleId={null}
    />
  </ErrorBoundary>
  <ErrorBoundary>
    <ArticlesDetails
      comments={Array []}
      onAddComment={[Function]}
      onAddReply={[Function]}
      onLoadReplies={[Function]}
      selectedArticle={undefined}
    />
  </ErrorBoundary>
</div>
`;

exports[`Home renders correctly while fetching 1`] = `
<div
  className="home"
>
  <ErrorBoundary>
    <Loading />
  </ErrorBoundary>
  <ErrorBoundary>
    <ArticlesList
      articles={Array []}
      articlesCount={0}
      onLoadComments={[Function]}
      onNextPage={[Function]}
      onPrevPage={[Function]}
      onSelectArticle={[Function]}
      page={0}
      selectedArticleId={null}
    />
  </ErrorBoundary>
  <ErrorBoundary>
    <ArticlesDetails
      comments={Array []}
      onAddComment={[Function]}
      onAddReply={[Function]}
      onLoadReplies={[Function]}
      selectedArticle={undefined}
    />
  </ErrorBoundary>
</div>
`;

exports[`Home renders correctly with params 1`] = `
<div
  className="home"
>
  <ErrorBoundary />
  <ErrorBoundary>
    <ArticlesList
      articles={Array []}
      articlesCount={0}
      onLoadComments={[Function]}
      onNextPage={[Function]}
      onPrevPage={[Function]}
      onSelectArticle={[Function]}
      page={0}
      selectedArticleId={null}
    />
  </ErrorBoundary>
  <ErrorBoundary>
    <ArticlesDetails
      comments={Array []}
      onAddComment={[Function]}
      onAddReply={[Function]}
      onLoadReplies={[Function]}
      selectedArticle={undefined}
    />
  </ErrorBoundary>
</div>
`;
